src/backend/distributed/planner/multi_logical_optimizer.c

```diff
@@ -2278,6 +2300,47 @@ TypeOid(Oid schemaId, const char *typeName)
 
 
 /*
+ * GroupedByDistinctPartitionColumn returns true if the query is grouped by the
+ * partition column of a table which has distinct partition values across shards.
+ */
+static bool
+GroupedByDistinctPartitionColumn(List *tableNodeList, MultiExtendedOp *opNode)
+{
+	bool groupedByDistinctPartitionColumn = false;
+	ListCell *tableNodeCell = NULL;
+
+	foreach(tableNodeCell, tableNodeList)
```

Comment:
Yeah, body of ``` if (partitionMethod == ...)``` inside the ```foreach``` looks almost identical in both functions. If we create ```GroupedByPartitionColumn()``` to reuse in both functions, we still need to keep most of what is inside ```TablePartitioningSupportsDistinct()``` as it is because of the ```/* if distinct is on table partition column, we can push it down */``` check, which I'm afraid will result in even more code. What do you think about keeping this refactoring until we have a 3rd example?

Reply:
Sure, I suppose leaving for a 3rd example is OK, but I worry no one will notice when we reach that. Leave it the way it is for now, I guess.